### PR TITLE
meteor: Improve specificity of types declared as "Object"

### PR DESCRIPTION
**Pull request by ebroder**
In all of the cases where we previously had a type declared as "Object", it was possible to use something more specific and better capture the actual constraints of the type.

This PR is a followup to a request from #64249. I have a [corresponding change](https://github.com/meteor/meteor/compare/release-2.11...ebroder:meteor:object-type-improvements?expand=1) prepared for the core Meteor type definitions, but I'm waiting to open the PR until we have confidence the changes are correct from the testing infrastructure here.

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

If changing an existing definition:
- [ ] Provide a URL to documentation or source code which provides context for the suggested changes: <<url here>>
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.

### CODE REVIEW

File: types/meteor/ejson.d.ts [hunk 1]
```diff
@@ -1,53 +1,46 @@
...
-    }
-    interface JSONable {
-        [key: string]: number | string | boolean | Object | number[] | string[] | Object[] | undefined | null;
-    }
-    interface EJSON extends EJSONable {}
+        | null
+        | EJSONable[]
+        | { [key: string]: EJSONable };
+
+    type JSONable = string | number | boolean | null | JSONable[] | { [key: string]: JSONable };
```
Comment (radekmie) [inline types/meteor/ejson.d.ts:20]: I'm not sure if that's a correct change (see https://github.com/microsoft/TypeScript/issues/1897 for details). Long story short, it doesn't handle custom serializers.

(Similarly, `fromJSONValue` has to return `any` _or_ have a type parameter to handle that case.)

I wouldn't touch these types just yet in this PR.

File: types/meteor/email.d.ts [hunk 2]
```diff
@@ -1,20 +1,10 @@
...
-        bcc?: string | string[] | undefined;
-        replyTo?: string | string[] | undefined;
-        subject?: string | undefined;
-        text?: string | undefined;
-        html?: string | undefined;
-        headers?: Object | undefined;
-        attachments?: Object[] | undefined;
-        mailComposer?: MailComposer | undefined;
-      }
+      type EmailOptions = { mailComposer: MailComposer } | SendMailOptions
```
Comment (radekmie) [inline types/meteor/email.d.ts:5]: `interface`s have one potential advantage - they can be extended in the application (thanks to declaration merging). Using `SendMailOptions` here is good idea, but it should remain an `interface`.

File: types/meteor/globals/reactive-dict.d.ts [hunk 3]
```diff
@@ -1,5 +1,5 @@
 
-declare class ReactiveDict<O = EJSONable> {
+declare class ReactiveDict<O = { [key: string]: EJSONable }> {
```
Comment (radekmie) [inline types/meteor/globals/reactive-dict.d.ts:2]: 1. As mentioned above, this should be a separate PR.
2. How about `Record<string, EJSONable>`?

File: types/meteor/meteor.d.ts [hunk 4]
```diff
@@ -248,7 +248,8 @@ declare module 'meteor/meteor' {
          * @param func A function that takes a callback as its final parameter
          * @param context Optional `this` object against which the original function will be invoked
          */
-        function wrapAsync(func: Function, context?: Object): any;
+        function wrapAsync<T extends Function>(func: T, context: ThisParameterType<T>): Function;
+        function wrapAsync(func: Function): Function;
```
Comment (radekmie) [inline types/meteor/meteor.d.ts:252]: The first line should be enough; maybe with making the `context` optional. _(It should remove the last callback argument, but this is already better than before.)_

File: types/meteor/meteor.d.ts [hunk 5]
```diff
@@ -302,7 +302,15 @@ declare module 'meteor/meteor' {
         ): void;
 
         function loginWithGoogle(
-            options?: Meteor.LoginWithExternalServiceOptions,
+            options?: Meteor.LoginWithExternalServiceOptions & {
+                // Google login accepts additional login parameters based on
+                // https://developers.google.com/identity/openid-connect/openid-connect#authenticationuriparameters.
+                // However, there's only one parameter that must be set directly; all
+                // others can be set using Meteor's standard OAuth login parameters
+                loginUrlParameters?: {
```
Comment (radekmie) [inline types/meteor/meteor.d.ts:310]: If you'd make that a JSDoc instead, it'll be accessible in the IDE tooltip.

File: types/meteor/meteor.d.ts [hunk 6]
```diff
@@ -436,15 +444,15 @@ declare module 'meteor/meteor' {
          * @param id The new document's ID.
          * @param fields The fields in the new document.  If `_id` is present it is ignored.
          */
-        added(collection: string, id: string, fields: Object): void;
+        added(collection: string, id: string, fields: Record<string, any>): void;
```
Comment (radekmie) [inline types/meteor/meteor.d.ts:447]: `unknown` would be a better choice here.

File: types/meteor/meteor.d.ts [hunk 7]
```diff
@@ -408,7 +416,7 @@ declare module 'meteor/meteor' {
             close: () => void;
             onClose: (callback: () => void) => void;
             clientAddress: string;
-            httpHeaders: Object;
+            httpHeaders: Record<string, string>;
```
Comment (radekmie) [inline types/meteor/meteor.d.ts:419]: Headers are most likely `string | string[]`.

**Review by radekmie — CHANGES_REQUESTED** [review]

File: types/meteor/email.d.ts (hunk 2 above)
Comment (ebroder) [reply types/meteor/email.d.ts:5]: Is it actually useful to be able to extend `EmailOptions`? `SendMailOptions` _is_ the type that Meteor accepts (it just passes it on to nodemailer), and this accurately captures things like the fact that `mailComposer` is mutually exclusive with any other options.

File: types/meteor/ejson.d.ts (hunk 1 above)
Comment (ebroder) [reply types/meteor/ejson.d.ts:20]: That's reasonable. What do you recommend using instead of `Object`, then? (Like, `Object` is generic enough that we might as well just call it `any`)

File: types/meteor/meteor.d.ts (hunk 4 above)
Comment (ebroder) [reply types/meteor/meteor.d.ts:252]: Ah yes, that's fair, I'll fix that. I originally tried to write a version that did drop the callback argument, but the semantics here are messy enough that I couldn't get it to work.

(It's not just that you drop the callback argument - the callback argument becomes optional, and the return type of the resulting function changes based on whether or not the callback is provided)

File: types/meteor/meteor.d.ts (hunk 6 above)
Comment (ebroder) [reply types/meteor/meteor.d.ts:447]: Did you mean `Record<string, unknown>` or just `unknown`? I'm pretty sure it _does_ have to be record type at minimum.

File: types/meteor/meteor.d.ts (hunk 7 above)
Comment (ebroder) [reply types/meteor/meteor.d.ts:419]: That's what I would have guessed, but this comes from sockjs, which declares them to just be string: https://github.com/DefinitelyTyped/DefinitelyTyped/blob/b483f710a8923b277896fc36de4454752a7d24b6/types/sockjs/index.d.ts#L39-L41

File: types/meteor/ejson.d.ts (hunk 1 above)
Comment (radekmie) [reply types/meteor/ejson.d.ts:20]: There's nothing really. I'd just leave it for now as `any` is far more permissive.

File: types/meteor/email.d.ts (hunk 2 above)
Comment (radekmie) [reply types/meteor/email.d.ts:5]: I don't know 🤷 If you suggest there's no need for that, I'm fine with that too.

File: types/meteor/meteor.d.ts (hunk 6 above)
Comment (radekmie) [reply types/meteor/meteor.d.ts:447]: `Record<string, unknown>`, sorry for being imprecise.

File: types/meteor/email.d.ts (hunk 2 above)
Comment (ebroder) [reply types/meteor/email.d.ts:5]: I guess there are two cases where it might be useful to add additional options - if you have a custom transport or if you register hook functions which mutate the email. I can add an additional interface that can be overridden and gets mixed into `EmailOptions`

**Comment by ebroder** [comment]
Alright. I think that covers everything you called out; thanks for the feedback. Feel free to take another look when you have a chance.

**Review by radekmie — APPROVED** [review]

**Comment by ebroder** [comment]
Ready to merge
